stan/math/rev/functor/integrate_1d.hpp

```diff
@@ -28,25 +28,42 @@ namespace math {
  * NaN, a std::domain_error is thrown
  *
  * @tparam F type of f
+ * @tparam Args types of arguments to f
+ * @param f function to compute gradients of
+ * @param x location at which to evaluate gradients
+ * @param xc complement of location (if bounded domain of integration)
+ * @param n compute gradient with respect to nth parameter
+ * @param msgs stream for messages
+ * @param args other arguments to pass to f
  */
-template <typename F>
+template <typename F, typename... Args>
 inline double gradient_of_f(const F &f, const double &x, const double &xc,
-                            const std::vector<double> &theta_vals,
-                            const std::vector<double> &x_r,
-                            const std::vector<int> &x_i, size_t n,
-                            std::ostream *msgs) {
+                            size_t n, std::ostream *msgs,
+                            const Args &... args) {
   double gradient = 0.0;
 
   // Run nested autodiff in this scope
   nested_rev_autodiff nested;
 
-  std::vector<var> theta_var(theta_vals.size());
-  for (size_t i = 0; i < theta_vals.size(); i++) {
-    theta_var[i] = theta_vals[i];
-  }
-  var fx = f(x, xc, theta_var, x_r, x_i, msgs);
+  std::tuple<decltype(deep_copy_vars(args))...> args_tuple_local_copy(
+      deep_copy_vars(args)...);
+
+  Eigen::VectorXd adjoints = Eigen::VectorXd::Zero(count_vars(args...));
+
+  var fx = apply(
+      [&f, &x, &xc, msgs](auto &&... args) { return f(x, xc, msgs, args...); },
+      args_tuple_local_copy);
+
   fx.grad();
-  gradient = theta_var[n].adj();
+
+  apply(
+      [&](auto &&... args) {
+        accumulate_adjoints(adjoints.data(),
+                            std::forward<decltype(args)>(args)...);
+      },
+      std::move(args_tuple_local_copy));
+
+  gradient = adjoints.coeff(n);
```

Comment:
I forgot to make this comment in my review, but optionally we can definitely figure out which arg has the adjoint value we need here in a clever way that doesn't require copying all of them. We could have some function to get the Nth var in a tuple but I'm also fine with not doing that in this PR

Reply:
We could have some internal function like 
```cpp
template <typename... Args>
double get_nth_adjoint(size_t n, const std::tuple<Args...>& tuple_arg) {
   size_t accum_vars = 0;
   bool stop_checking = false;
   // for_each goes off from left to right
   std::array<double, sizeof...(Args)> possible_adjs = for_each([&accum_vars, &stop_checking](auto&& arg){
      if (stop_checking) return 0.0;
      size_t num_vars = count_vars(arg);
      // Need to keep moving along
      if ((accum_vars + num_vars) < nth || stop_checking) {
         accum_vars += num_vars;
         return 0.0;
      } else { // We reached the first arg that passes
         stop_checking = true;
         // I'm tired but do the logic here to get the nth value from that particular arg
         return the_adj;
      } 
   }, tuple_arg);
   // Loop over possible_adjs until we hit a nonzero value (or they are all zero and return a zero)
};
```

Reply:
Yeah it's awkward. I also want to leave it for now.

The way to speed this up is writing our own quadratures or talking to the Boost people about an interface where we integrate multiple functions on the same domain together. Now we compute all three of these integrals totally separately:

```
\int f(x, a, b) dx
\int df(x, a, b)/da
\int df(x, a, b)/db
```

But anytime we compute df(x, a, b)/da we also get df(x, a, b)/db, and so the efficiency gains would be taking advantage of that (and what we're doing here is throwing away a ton of gradient info).